### PR TITLE
PDS uses sechub storage + enhanced DAUI

### LINKED ISSUE #435: Provide  "use sechub storage" possibility for PDS
When #434 is done we should introduce a "reuse" option for PDS servers, so they can directly reuse NFS or S3 storage from sechub to get job data (e.g source zip file) without extra upload!

Makes PDS actions much more faster and reduces traffic!

---
<sup>Albert Tregnaghi <albert.tregnaghi@daimler.com>, Daimler TSS GmbH, [imprint](https://github.com/Daimler/daimler-foss/blob/master/LEGAL_IMPRINT.md)</sup>

### LINKED ISSUE #646: Key value parameter dialog shall show complete descriptions
## Situation

When we define a new executor configuration - e.g. for PDS - the description part for paramters is often not full rendered and shown.

## Wanted
Improve output for key value parameters. Type, Necessarity and description shall  always be readable.
Also some highlighting shall be applied to them.

### PR DESCRIPTION
**Pull request by de-jcup**
This PR 

- closes #646 
- closes #435

### CODE REVIEW

**Review by Jeeppler — CHANGES_REQUESTED** [review]
See comments.

File: sechub-adapter-pds/src/test/java/com/daimler/sechub/adapter/pds/PDSAdapterV1WireMockTest.java [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+    private Map<String, String> expectedJobParameters;
+
+    @Before
+    public void beforeEach() {
+        adapterToTest = new PDSAdapterV1();
+        callback = mock(AdapterMetaDataCallback.class);
+
+        productIdentifier = "EXAMPLE_PRODUCT";
+        sechubJobUUID = UUID.randomUUID();
+        expectedJobParameters = new TreeMap<>();
```
[the remark this answers is not in the document]
Comment (de-jcup) [reply sechub-adapter-pds/src/test/java/com/daimler/sechub/adapter/pds/PDSAdapterV1WireMockTest.java:51]: The reason for this is that a tree map is always sorted. this is easier to debug. And we are talking about some simple parameters.

File: sechub-adapter-pds/src/test/java/com/daimler/sechub/adapter/pds/PDSWiremockTestSupport.java [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+        PDSAdapterJobStatusState state;
+    }
+
+    private static class GetResult {
+        String result;
+        int exitCode;
+        boolean failed;
+    }
+
+//    public IcrementalAdditionalPrefixAPIURLSupport getApiURLSupport() {
```
[the remark this answers is not in the document]
Comment (de-jcup) [reply sechub-adapter-pds/src/test/java/com/daimler/sechub/adapter/pds/PDSWiremockTestSupport.java:60]: dropped comments

File: sechub-adapter-pds/src/test/java/com/daimler/sechub/adapter/pds/PDSWiremockTestSupport.java [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+        return new PDSWiremockTestSupportBuilder(server);
+    }
+
+    public void startPDSServerSimulation() {
+
+        try {
+            if (pdsJobCanBeCreated) {
+                simulationJobCreation();
+            }
+            for (UploadInfo info : uploads) {
```
[the remark this answers is not in the document]
Comment (de-jcup) [reply sechub-adapter-pds/src/test/java/com/daimler/sechub/adapter/pds/PDSWiremockTestSupport.java:74]: At the moment we only use sourcecode.zip - but we will provide more in future - e.g. binary-arifact.obj or something.
The accepted uploads are managed by PDS server. The test support here just provides a possiblity to handle this in our tests as well - e.g. when somebody tries to upload sourcecode.zip files again and again etc. (another testscenario, but currently not implemented).

File: sechub-adapter-pds/src/test/java/com/daimler/sechub/adapter/pds/PDSWiremockTestSupport.java [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+            for (StateQueueInfo info : stateRequestsResults) {
+                simulateStateRequest(info);
+            }
+            if (pdsFetchJobResult != null) {
+                simulateGetResult(pdsFetchJobResult);
+            }
+
+        } catch (
+
+        Exception e) {
```
[the remark this answers is not in the document]
Comment (de-jcup) [reply sechub-adapter-pds/src/test/java/com/daimler/sechub/adapter/pds/PDSWiremockTestSupport.java:89]: done

File: sechub-adapter-pds/src/test/java/com/daimler/sechub/adapter/pds/PDSWiremockTestSupport.java [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+
+    private void simulateUploadData(String uploadFileName) {
+
+        if (pdsJobUUID == null) {
+            throw new IllegalStateException("testcase corrupt? pds job uuid not known here!");
+        }
+        /* @formatter:off */
+        String url = pdsURLBuilder.buildUpload(pdsJobUUID, uploadFileName);
+        stubFor(post(urlEqualTo(history.rememberPOST(url)))
+                // .inScenario(chain.getScenario()).whenScenarioStateIs(chain.getStateBefore())
```
[the remark this answers is not in the document]
Comment (de-jcup) [reply sechub-adapter-pds/src/test/java/com/daimler/sechub/adapter/pds/PDSWiremockTestSupport.java:147]: done

File: sechub-developertools/src/main/java/com/daimler/sechub/developertools/admin/ui/action/adapter/ProductExecutorTemplatesDialogUI.java [hunk 6]
```diff
@@ -95,7 +95,11 @@ public TemplatesDialogResult showDialog(TemplatesDialogConfig config) {
         JMenuBar menuBar = createMainMenu();
 
         if (config.inputContent != null) {
+            /* we use the importer all time to initialize content */
             SwingUtilities.invokeLater(() -> importFromClipboardAction.importText(config.inputContent));
+        }else {
+            /* when new at all */
+            System.out.println("No input content found!!!");
```
[the remark this answers is not in the document]
Comment (de-jcup) [reply sechub-developertools/src/main/java/com/daimler/sechub/developertools/admin/ui/action/adapter/ProductExecutorTemplatesDialogUI.java:102]: Maybe, maybe not ;-) - but I removed it as suggested. So done.

File: sechub-doc/src/docs/asciidoc/diagrams/diagram_concept_pds_storage_example_shared_s3_bucket.puml [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+            component scanJob1 as "Scan job"
+            storage localFS1 as "Local filesystem"
+        }
+        
+    }
+    database secHubDB as "SecHub database" {
+            
+            rectangle XYZConfig as "Product executor config XYZ"
+    }
+
```
[the remark this answers is not in the document]
Comment (de-jcup) [reply sechub-doc/src/docs/asciidoc/diagrams/diagram_concept_pds_storage_example_shared_s3_bucket.puml:39]: done

File: sechub-doc/src/docs/asciidoc/documents/shared/concepts/concept_pds.adoc [hunk 8]
```diff
@@ -32,6 +32,135 @@ need to provide a `ProductDelegationServer` (in short form `PDS`).
 ==== Big picture
 plantuml::diagrams/diagram_concept_product_delgation_server_bigpicture.puml[]
 
+
+[[pds-storage-and-sharing]]
+==== Storage and sharing
+
+{pds} needs - like {sechub} - the possiblity to store job data in a central way when operating
```
[the remark this answers is not in the document]
Comment (de-jcup) [reply sechub-doc/src/docs/asciidoc/documents/shared/concepts/concept_pds.adoc:39]: Done

File: sechub-doc/src/docs/asciidoc/documents/shared/concepts/concept_pds.adoc [hunk 9]
```diff
@@ -32,6 +32,135 @@ need to provide a `ProductDelegationServer` (in short form `PDS`).
...
+inside a cluster _(it's not clear which cluster member uploads job data and which one does execute 
+the job and need the data at exectuion time)._
+
+For {secHub} there were already existing (and autark) storage mechanism for S3 and for shared volumes.
```
[the remark this answers is not in the document]
Comment (de-jcup) [reply sechub-doc/src/docs/asciidoc/documents/shared/concepts/concept_pds.adoc:43]: Changed to
```asciidoc
{pds} does use the `sechub-storage-*` libraries which provide storage mechanism for S3 and 
for shared volumes.
```
So it becomes more clear.

File: sechub-doc/src/docs/asciidoc/documents/shared/concepts/concept_pds.adoc [hunk 10]
```diff
@@ -32,6 +32,135 @@ need to provide a `ProductDelegationServer` (in short form `PDS`).
...
+
+[TIP]
+====
+The next examples explain the different situations appearing for a PDS cluster executing product "XYZ", 
+but of course this applies to multiple clusters for different products as well
+====
+
+[NOTE]
+====
+For non clustered environment "shared volumes" can also be normal local file system pathes like
```
[the remark this answers is not in the document]
Comment (de-jcup) [reply sechub-doc/src/docs/asciidoc/documents/shared/concepts/concept_pds.adoc:53]: done

File: sechub-doc/src/docs/asciidoc/documents/shared/concepts/concept_pds.adoc [hunk 11]
```diff
@@ -32,6 +32,135 @@ need to provide a `ProductDelegationServer` (in short form `PDS`).
...
+When talking about a NFS or a shared volume here, this is always meant to be a file system path.
+It's important to mount a NFS at a local file path for the PDS server when using a shared volume.
+URLs are NOT supported - only filepathes. So you must mount your network file storage to a local file location!
+
+For example: a base path for a shared volume could look like: `/media/mounted-storage/`
+====
+
+
+===== Shared S3 storage
+In the next example PDS and SecHub are using same S3 bucket as storage.
```
[the remark this answers is not in the document]
Comment (de-jcup) [reply sechub-doc/src/docs/asciidoc/documents/shared/concepts/concept_pds.adoc:68]: done

File: sechub-doc/src/docs/asciidoc/documents/shared/concepts/concept_pds.adoc [hunk 12]
```diff
@@ -32,6 +32,135 @@ need to provide a `ProductDelegationServer` (in short form `PDS`).
...
+
+We have two different scenarios here:
+
+* Product executor configuration `pds.config.use.sechub.storage` is `true` +
+  In this case the way is exactly as shown in next picture: +
```
[the remark this answers is not in the document]
Comment (de-jcup) [reply sechub-doc/src/docs/asciidoc/documents/shared/concepts/concept_pds.adoc:73]: done

File: sechub-doc/src/docs/asciidoc/documents/shared/concepts/concept_pds.adoc [hunk 13]
```diff
@@ -32,6 +32,135 @@ need to provide a `ProductDelegationServer` (in short form `PDS`).
...
+  
+[[pds-storage-different-s3]]
+===== Different S3 storages
+In the next example PDS and SecHub are using different S3 buckets as storage.
+
+We have two different scenarios here:
+
+* Product executor configuration `pds.config.use.sechub.storage` is NOT `true` +
+  In this case the way is exactly as shown in next picture: +
+  - The user uploads job data for a  {secHub} job and it is stored at shared S3 bucket _(blue lines)_.
```
[the remark this answers is not in the document]
Comment (de-jcup) [reply sechub-doc/src/docs/asciidoc/documents/shared/concepts/concept_pds.adoc:95]: done

File: sechub-doc/src/docs/asciidoc/documents/shared/concepts/concept_pds.adoc [hunk 14]
```diff
@@ -32,6 +32,135 @@ need to provide a `ProductDelegationServer` (in short form `PDS`).
...
+WARNING: This will NOT WORK! The job storage will not be found and a error will happen at job execution time.
+ 
+===== Same shared volume (NFS)
+In the next example PDS server and SecHub are using same shared volume as storage.
+
+We have two different scenarios here:
+
+* Product executor configuration `pds.config.use.sechub.storage` is `true` +
+  In this case the way is exactly as shown in next picture: +
+  - The user uploads job data for a  {secHub} job and it is stored at shared volume _(blue lines)_.
```
[the remark this answers is not in the document]
Comment (de-jcup) [reply sechub-doc/src/docs/asciidoc/documents/shared/concepts/concept_pds.adoc:117]: done

File: sechub-doc/src/docs/asciidoc/documents/shared/concepts/concept_pds.adoc [hunk 15]
```diff
@@ -32,6 +32,135 @@ need to provide a `ProductDelegationServer` (in short form `PDS`).
...
+  - When {secHub} job is starting it will create a {pds} job but does not upload any additional data. 
+  - When PDS job is starting the  PDS server will fetch workspace files from existing {secHub} job data _(red lines)_.
```
[the remark this answers is not in the document]
Comment (de-jcup) [reply sechub-doc/src/docs/asciidoc/documents/shared/concepts/concept_pds.adoc:119]: done

File: sechub-doc/src/docs/asciidoc/documents/shared/concepts/concept_pds.adoc [hunk 16]
```diff
@@ -32,6 +32,135 @@ need to provide a `ProductDelegationServer` (in short form `PDS`).
...
+  
+[[pds-storage-different-sharedvolume]]
+===== Different shared volumes (NFS)
+In the next example PDS and SecHub are using different shared volumes as storage.
+
+We have two different scenarios here:
+
+* Product executor configuration `pds.config.use.sechub.storage` is NOT `true` +
+  In this case the way is exactly as shown in next picture: +
+  - The user uploads job data for a  {secHub} job and it is stored at sechub shared volume _(blue lines)_.
```
[the remark this answers is not in the document]
Comment (de-jcup) [reply sechub-doc/src/docs/asciidoc/documents/shared/concepts/concept_pds.adoc:138]: done

File: sechub-doc/src/docs/asciidoc/documents/shared/pds_storage_configuration.adoc [hunk 17]
```diff
@@ -4,4 +4,27 @@ In PDS we need to store job data (e.g. zipped source code).
 *At least one storage setup must be defined*- otherwise {pds} server will not start!
 You can either define a shared volume (normally a NFS) or a S3 storage.
 
-Look at <<section-gen-config-scope-storage,Storage configuration>> for configuration details.
\ No newline at end of file
+Look at <<section-gen-config-scope-storage,Storage configuration>> for configuration details.
+
+==== Usage of SecHub storage
+
+When using own storage for PDS servers, {secHub}  has to upload already pesisted job data _(e.g. zipped sourcecode for SAST scannings)_
```
[the remark this answers is not in the document]
Comment (de-jcup) [reply sechub-doc/src/docs/asciidoc/documents/shared/pds_storage_configuration.adoc:11]: done

File: sechub-pds/src/test/java/com/daimler/sechub/pds/job/PDSWorkspaceServiceTest.java [hunk 18]
```diff
@@ -3,42 +3,89 @@
...
+        /* execute */
+        serviceToTest.prepareWorkspace(jobUUID, config);
+
+        /* test */
+        verify(storageService).getJobStorage("xyz/abc/project1",secHubJobUUID);
+    }
+    
+    private PDSExecutionParameterEntry createEntry(String key, String value) {
+        PDSExecutionParameterEntry entry = new PDSExecutionParameterEntry();
+        entry.setKey(key);;
```
[the remark this answers is not in the document]
Comment (de-jcup) [reply sechub-pds/src/test/java/com/daimler/sechub/pds/job/PDSWorkspaceServiceTest.java:86]: done

File: sechub-pds/src/test/java/com/daimler/sechub/pds/job/PDSWorkspaceServiceTest.java [hunk 19]
```diff
@@ -3,42 +3,89 @@
...
+        entry.setValue(value);;
```
[the remark this answers is not in the document]
Comment (de-jcup) [reply sechub-pds/src/test/java/com/daimler/sechub/pds/job/PDSWorkspaceServiceTest.java:87]: done

File: sechub-testframework/src/main/java/com/daimler/sechub/test/JSONTestUtil.java [hunk 20]
```diff
@@ -29,6 +29,30 @@ public static String toJSONContainingNullValues(List<Map<String, Object>> list)
...
+     * this).
+     * 
+     * @param map
+     * @return JSON string
+     */
+    public static String toJSONContainingAnything(List<Object> list) {
+        StringBuilder sb = new StringBuilder();
+        
+        sb.append("[");
+        for (Iterator<Object> it = list.iterator();it.hasNext();) {
```
[the remark this answers is not in the document]
Comment (de-jcup) [reply sechub-testframework/src/main/java/com/daimler/sechub/test/JSONTestUtil.java:44]: Still valid - same as in another review:

```java
           if (it.hasNext()) {
                sb.append(",\n");
            }
```
only possible on iterators....

So will keep as is.

File: sechub-doc/src/docs/asciidoc/documents/shared/concepts/concept_pds.adoc [hunk 21]
```diff
@@ -32,6 +32,135 @@ need to provide a `ProductDelegationServer` (in short form `PDS`).
...
+  - When PDS job is starting, the  PDS server will fetch workspace files from its {pds} job data _(red lines)_. 
+  - The storage path location for {pds} is `${pdsSharedVolumeBasePath}/pds/${pdsProductIdentifier}/${pdsJobUUID}/` 
+ 
+ 
+plantuml::diagrams/diagram_concept_pds_storage_example_different_nfs.puml[]
+
+
+* Product executor configuration `pds.config.use.sechub.storage` is `true`
+
+WARNING: This will NOT WORK! The job storage will not be found and a error will happen at job execution time.
```
[the remark this answers is not in the document]
Comment (de-jcup) [reply sechub-doc/src/docs/asciidoc/documents/shared/concepts/concept_pds.adoc:151]: done

File: sechub-doc/src/docs/asciidoc/documents/shared/concepts/concept_pds.adoc [hunk 22]
```diff
@@ -32,6 +32,135 @@ need to provide a `ProductDelegationServer` (in short form `PDS`).
...
+  - When PDS job is starting, the  PDS server will fetch workspace files from its {pds} job data _(red lines)_. 
+  - The storage path location for {pds} is `${pdsS3Bucket}/pds/${pdsProductIdentifier}/${pdsJobUUID}/` 
+
+
+plantuml::diagrams/diagram_concept_pds_storage_example_different_s3_buckets.puml[]
+
+
+* Product executor configuration `pds.config.use.sechub.storage` is `true`
+
+WARNING: This will NOT WORK! The job storage will not be found and a error will happen at job execution time.
```
[the remark this answers is not in the document]
Comment (de-jcup) [reply sechub-doc/src/docs/asciidoc/documents/shared/concepts/concept_pds.adoc:108]: done

**Review by Jeeppler — APPROVED** [review]

**Pull request merged**